File: 2022/day15/perl/solution.pl
#!/usr/bin/env perl
use strict;
use warnings;
use File::Basename;
use Cwd 'abs_path';

# Parse sensor and beacon positions from input
sub parse_sensors {
    my ($text) = @_;
    my @sensors;

    for my $line (split /\n/, $text) {
        next unless $line =~ /Sensor at x=(-?\d+), y=(-?\d+): closest beacon is at x=(-?\d+), y=(-?\d+)/;
        my ($sx, $sy, $bx, $by) = ($1, $2, $3, $4);
        my $dist = abs($sx - $bx) + abs($sy - $by);  # Manhattan distance
        push @sensors, [$sx, $sy, $bx, $by, $dist];
    }

    return \@sensors;
}

# Merge overlapping ranges
sub merge_ranges {
    my ($ranges) = @_;
    return [] unless @$ranges;

    my @sorted = sort { $a->[0] <=> $b->[0] || $a->[1] <=> $b->[1] } @$ranges;
    my @merged = ($sorted[0]);

    for my $i (1 .. $#sorted) {
        my ($start, $end) = @{$sorted[$i]};
        if ($start <= $merged[-1][1] + 1) {
            $merged[-1][1] = $end if $end > $merged[-1][1];
        } else {
            push @merged, [$start, $end];
        }
    }

    return \@merged;
}

# Get ranges covered by sensors at a specific row
sub get_coverage_at_row {
    my ($sensors, $row) = @_;
    my @ranges;

    for my $sensor (@$sensors) {
        my ($sx, $sy, $bx, $by, $dist) = @$sensor;
        my $row_dist = abs($sy - $row);
        next if $row_dist > $dist;  # Sensor doesn't reach this row

        my $x_spread = $dist - $row_dist;
        push @ranges, [$sx - $x_spread, $sx + $x_spread];
    }

    return merge_ranges(\@ranges);
}

# Part 1: Count positions that cannot contain a beacon at row y=2000000
sub part1 {
    my ($sensors) = @_;
    my $target_row = 2000000;

    my $ranges = get_coverage_at_row($sensors, $target_row);

    # Count total coverage
    my $total = 0;
    for my $range (@$ranges) {
        $total += $range->[1] - $range->[0] + 1;
    }

    # Subtract beacons that are on this row
    my %beacons_on_row;
    for my $sensor (@$sensors) {
        my ($sx, $sy, $bx, $by, $dist) = @$sensor;
        if ($by == $target_row) {
            $beacons_on_row{$bx} = 1;
        }
    }

    return $total - scalar(keys %beacons_on_row);
}

# Part 2: Find the distress beacon's tuning frequency
sub part2 {
    my ($sensors) = @_;
    my $max_coord = 4000000;

    for my $row (0 .. $max_coord) {
        my $ranges = get_coverage_at_row($sensors, $row);

        # Clip ranges to search area
        my @clipped;
        for my $range (@$ranges) {
            my ($start, $end) = @$range;
            next if $end < 0 || $start > $max_coord;
            push @clipped, [
                $start < 0 ? 0 : $start,
                $end > $max_coord ? $max_coord : $end
            ];
        }

        my $merged = merge_ranges(\@clipped);

        # Check if full row is covered
        if (@$merged == 1 && $merged->[0][0] == 0 && $merged->[0][1] == $max_coord) {
            next;
        }

        # Found a gap - the beacon is in the gap
        my $x;
        if (@$merged > 1) {
            # Gap between ranges
            $x = $merged->[0][1] + 1;
        } elsif ($merged->[0][0] > 0) {
            $x = 0;
        } else {
            $x = $merged->[0][1] + 1;
        }

        return $x * 4000000 + $row;
    }

    return undef;
}

sub main {
    my $script_dir = dirname(abs_path($0));
    my $input_file = "$script_dir/../input.txt";

    open my $fh, '<', $input_file or die "Cannot open $input_file: $!";
    my $text = do { local $/; <$fh> };
    close $fh;

    my $sensors = parse_sensors($text);

    print "Part 1: ", part1($sensors), "\n";
    print "Part 2: ", part2($sensors), "\n";
}

main();

File: 2022/day15/python/solution.py
#!/usr/bin/env python3
import os
import re

def parse_sensors(text):
    """Parse sensor and beacon positions."""
    sensors = []
    pattern = r'Sensor at x=(-?\d+), y=(-?\d+): closest beacon is at x=(-?\d+), y=(-?\d+)'

    for line in text.strip().split('\n'):
        match = re.match(pattern, line)
        sx, sy, bx, by = map(int, match.groups())
        dist = abs(sx - bx) + abs(sy - by)  # Manhattan distance
        sensors.append((sx, sy, bx, by, dist))

    return sensors

def get_coverage_at_row(sensors, row):
    """Get ranges covered by sensors at a specific row."""
    ranges = []

    for sx, sy, bx, by, dist in sensors:
        # How far is the row from the sensor?
        row_dist = abs(sy - row)
        if row_dist > dist:
            continue  # Sensor doesn't reach this row

        # Calculate x-range covered at this row
        x_spread = dist - row_dist
        ranges.append((sx - x_spread, sx + x_spread))

    return merge_ranges(ranges)

def merge_ranges(ranges):
    """Merge overlapping ranges."""
    if not ranges:
        return []

    ranges.sort()
    merged = [ranges[0]]

    for start, end in ranges[1:]:
        if start <= merged[-1][1] + 1:
            merged[-1] = (merged[-1][0], max(merged[-1][1], end))
        else:
            merged.append((start, end))

    return merged

def part1(text):
    """Count positions that cannot contain a beacon at row y=2000000."""
    sensors = parse_sensors(text)
    target_row = 2000000

    ranges = get_coverage_at_row(sensors, target_row)

    # Count total coverage
    total = sum(end - start + 1 for start, end in ranges)

    # Subtract beacons that are on this row
    beacons_on_row = set()
    for sx, sy, bx, by, dist in sensors:
        if by == target_row:
            beacons_on_row.add(bx)

    return total - len(beacons_on_row)

def part2(text):
    """Find the distress beacon's tuning frequency."""
    sensors = parse_sensors(text)
    max_coord = 4000000

    # For each row, find if there's a gap in coverage
    for row in range(max_coord + 1):
        ranges = get_coverage_at_row(sensors, row)

        # Clip ranges to search area
        clipped = []
        for start, end in ranges:
            if end < 0 or start > max_coord:
                continue
            clipped.append((max(0, start), min(max_coord, end)))

        clipped = merge_ranges(clipped)

        # Check if full row is covered
        if len(clipped) == 1 and clipped[0][0] == 0 and clipped[0][1] == max_coord:
            continue

        # Found a gap - the beacon is in the gap
        if len(clipped) > 1:
            # Gap between ranges
            x = clipped[0][1] + 1
        elif clipped[0][0] > 0:
            x = 0
        else:
            x = clipped[0][1] + 1

        return x * 4000000 + row

    return None

def main():
    script_dir = os.path.dirname(os.path.abspath(__file__))
    input_file = os.path.join(script_dir, '..', 'input.txt')

    with open(input_file) as f:
        text = f.read()

    print('Part 1:', part1(text))
    print('Part 2:', part2(text))

if __name__ == '__main__':
    main()
